src/main/java/org/opensearch/ad/rest/RestExecuteAnomalyDetectorAction.java

```diff
@@ -19,73 +19,104 @@
 
-    public RestExecuteAnomalyDetectorAction(Settings settings, ClusterService clusterService) {
-        this.requestTimeout = REQUEST_TIMEOUT.get(settings);
-        clusterService.getClusterSettings().addSettingsUpdateConsumer(REQUEST_TIMEOUT, it -> requestTimeout = it);
+    public RestExecuteAnomalyDetectorAction(ExtensionsRunner extensionsRunner, SDKRestClient sdkRestClient) {
+        this.requestTimeout = REQUEST_TIMEOUT.get(extensionsRunner.getEnvironmentSettings());
+        this.sdkRestClient = sdkRestClient;
+        extensionsRunner.getSdkClusterService().getClusterSettings().addSettingsUpdateConsumer(REQUEST_TIMEOUT, it -> requestTimeout = it);
+        this.settings = extensionsRunner.getEnvironmentSettings();
     }
 
-    @Override
     public String getName() {
         return DETECT_DATA_ACTION;
     }
 
-    @Override
-    protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient client) throws IOException {
+    protected ExtensionRestResponse prepareRequest(RestRequest request) throws IOException {
         if (!EnabledSetting.isADPluginEnabled()) {
             throw new IllegalStateException(CommonErrorMessages.DISABLED_ERR_MSG);
         }
         AnomalyDetectorExecutionInput input = getAnomalyDetectorExecutionInput(request);
-        return channel -> {
-            String rawPath = request.rawPath();
-            String error = validateAdExecutionInput(input);
-            if (StringUtils.isNotBlank(error)) {
-                channel.sendResponse(new BytesRestResponse(RestStatus.BAD_REQUEST, error));
-                return;
-            }
-
-            AnomalyResultRequest getRequest = new AnomalyResultRequest(
-                input.getDetectorId(),
-                input.getPeriodStart().toEpochMilli(),
-                input.getPeriodEnd().toEpochMilli()
+
+        String rawPath = request.rawPath();
+        String error = validateAdExecutionInput(input);
+
+        if (StringUtils.isNotBlank(error)) {
+            return getExecuteDetectorResponse(RestStatus.BAD_REQUEST, request, null, error);
+        }
+
+        CompletableFuture<AnomalyResultResponse> futureResponse = new CompletableFuture<>();
+
+        AnomalyResultRequest anomalyResultRequest = new AnomalyResultRequest(
+            input.getDetectorId(),
+            input.getPeriodStart().toEpochMilli(),
+            input.getPeriodEnd().toEpochMilli()
+        );
+
+        sdkRestClient
+            .execute(
+                AnomalyResultAction.INSTANCE,
+                anomalyResultRequest,
+                ActionListener.wrap(r -> futureResponse.complete(r), e -> futureResponse.completeExceptionally(e))
             );
-            client.execute(AnomalyResultAction.INSTANCE, getRequest, new RestToXContentListener<>(channel));
-        };
+
+        AnomalyResultResponse anomalyResultResponse = futureResponse
+            .orTimeout(AnomalyDetectorSettings.REQUEST_TIMEOUT.get(settings).getMillis(), TimeUnit.MILLISECONDS)
```

Comment:
Since we have `this.requestTimeout` already defined in the constructor. Fetch mills directly from the variable:
```suggestion
            .orTimeout(this.requestTimeout.getMillis(), TimeUnit.MILLISECONDS)
```